--- a/ggml/src/ggml-cuda/conv-transpose-1d.cu
+++ b/ggml/src/ggml-cuda/conv-transpose-1d.cu
@@ -1,51 +1,53 @@
 #include "conv-transpose-1d.cuh"
+#include <algorithm>
 
 static  __global__ void conv_transpose_1d_kernel(
-        const int s0, const int p0, const int d0, const int output_size,
+        const int s0, const int p0, const int d0, const int64_t output_size,
         const int src0_ne0, const int src0_ne1, const int src0_ne2, const int src0_ne3,
         const int src1_ne0, const int src1_ne1, const int src1_ne2, const int src1_ne3,
         const int dst_ne0, const int dst_ne1, const int dst_ne2, const int dst_ne3,
         const float * src0, const float * src1,  float * dst) {
-    int global_index = threadIdx.x + blockIdx.x * blockDim.x;
-    if (global_index >= output_size) {
-        return;
-    }
 
-    int out_index = global_index / dst_ne0;
+    for (int64_t global_index = threadIdx.x + blockIdx.x * blockDim.x; global_index < output_size; global_index += blockDim.x * gridDim.x) {
+        int out_index = global_index / dst_ne0;
 
-    float accumulator = 0;
+        float accumulator = 0;
 
-    for (int c = 0; c < src0_ne2; c++) {
-        int idx = global_index % dst_ne0;
+        for (int c = 0; c < src0_ne2; c++) {
+            int idx = global_index % dst_ne0;
 
-        int kernel_offset = (src0_ne0 * src0_ne1 * c) + (out_index * src0_ne0);
-        int input_offset = src1_ne0 * c;
+            int kernel_offset = (src0_ne0 * src0_ne1 * c) + (out_index * src0_ne0);
+            int input_offset = src1_ne0 * c;
 
-        for (int i = 0; i < src1_ne0; i++) {
-            if (!(idx >= i*s0 && idx < i*s0 + src0_ne0)) {
-                continue;
-            }
-            int weight_idx = idx - i*s0;
 
-            float kernel_weight = src0[kernel_offset + weight_idx];
-            float input_value =  src1[input_offset+i];
+            int start_i = (idx >= src0_ne0) ? (idx - src0_ne0 + s0) / s0 : 0;
+            int end_i = idx / s0;
+
+            if (end_i >= src1_ne0) end_i = src1_ne0 - 1;
 
-            accumulator += kernel_weight * input_value;
+            for (int i = start_i; i <= end_i; i++) {
+                int weight_idx = idx - i*s0;
+                float kernel_weight = src0[kernel_offset + weight_idx];
+                float input_value =  src1[input_offset+i];
+
+                accumulator += kernel_weight * input_value;
+            }
         }
+        dst[global_index] = accumulator;
     }
-    dst[global_index] = accumulator;
     GGML_UNUSED_VARS(p0, d0, src0_ne3, src1_ne3, dst_ne3, src1_ne1, dst_ne1, src1_ne2, dst_ne2);
 }
 
 static void conv_transpose_1d_f32_f32_cuda(
-        const int s0, const int p0, const int d0, const int output_size,
+        const int s0, const int p0, const int d0, const int64_t output_size,
         const int src0_ne0, const int src0_ne1, const int src0_ne2, const int src0_ne3,
         const int src1_ne0, const int src1_ne1, const int src1_ne2, const int src1_ne3,
         const int dst_ne0, const int dst_ne1, const int dst_ne2, const int dst_ne3,
         const float * src0, const float * src1,  float * dst,
         cudaStream_t stream) {
 
-    const int num_blocks = (output_size + CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE - 1) / CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE;
+    const int num_blocks = std::min((int64_t)65535, (output_size + CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE - 1) / CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE);
+    
     conv_transpose_1d_kernel<<<num_blocks,CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE, 0, stream>>>(
         s0,p0,d0,output_size,
         src0_ne0, src0_ne1,  src0_ne2, src0_ne3,
@@ -83,4 +85,4 @@ void ggml_cuda_op_conv_transpose_1d(ggml_backend_cuda_context & ctx, ggml_tensor
         src1->ne[0], src1->ne[1], src1->ne[2], src1->ne[3],
         dst->ne[0], dst->ne[1], dst->ne[2], dst->ne[3],
         src0_d, src1_d, dst_d, stream);
-}
+}
\ No newline at end of file
